Cupertino transparent navigation bars

veloce opened the pull request:
This PR is making the `CupertinoNavigationBar` and `CupertinoSliverNavigationBar` appear transparent as long as the content is not scrolled under them, so they look like standard iOS apps nav bars.

https://github.com/flutter/flutter/assets/423393/eee2700b-2a91-4577-922c-6163d47cb357

https://github.com/flutter/flutter/assets/423393/3847f2b5-0dac-4d5e-aa6f-03c1d2893e30

<details>
  <summary>Demo app code</summary>
  
  ```dart
import 'package:flutter/cupertino.dart';

/// Flutter code sample for [CupertinoTabScaffold].

void main() => runApp(const TabScaffoldApp());

class TabScaffoldApp extends StatefulWidget {
  const TabScaffoldApp({super.key});

  @override
  State<TabScaffoldApp> createState() => _TabScaffoldAppState();
}

class _TabScaffoldAppState extends State<TabScaffoldApp> {
  Brightness _brightness = Brightness.light;

  @override
  Widget build(BuildContext context) {
    return CupertinoApp(
      theme: CupertinoThemeData(brightness: _brightness),
      home: TabScaffoldExample(
          brightness: _brightness, onBrightnessToggle: _toggleBrightness),
    );
  }

  void _toggleBrightness() {
    setState(() {
      _brightness =
          _brightness == Brightness.light ? Brightness.dark : Brightness.light;
    });
  }
}

class TabScaffoldExample extends StatefulWidget {
  const TabScaffoldExample(
      {required this.brightness, required this.onBrightnessToggle, super.key});

  final VoidCallback onBrightnessToggle;
  final Brightness brightness;

  @override
  State<TabScaffoldExample> createState() => _TabScaffoldExampleState();
}

class _TabScaffoldExampleState extends State<TabScaffoldExample> {
  @override
  Widget build(BuildContext context) {
    return CupertinoTabScaffold(
      tabBar: CupertinoTabBar(
        items: const <BottomNavigationBarItem>[
          BottomNavigationBarItem(
            icon: Icon(CupertinoIcons.home),
            label: 'Home',
          ),
          BottomNavigationBarItem(
            icon: Icon(CupertinoIcons.search_circle_fill),
            label: 'Explore',
          ),
          BottomNavigationBarItem(
            icon: Icon(CupertinoIcons.person_fill),
            label: 'Profile',
          ),
          BottomNavigationBarItem(
            icon: Icon(CupertinoIcons.settings_solid),
            label: 'Settings',
          ),
        ],
      ),
      tabBuilder: (BuildContext context, int index) {
        return CupertinoTabView(
          builder: (BuildContext context) {
            return CupertinoPageScaffold(
              backgroundColor: index == 3
                  ? CupertinoColors.secondarySystemBackground
                      .resolveFrom(context)
                  : null,
              child: CustomScrollView(
                slivers: [
                  CupertinoSliverNavigationBar(
                    largeTitle: Text('Tab $index'),
                    initiallyTransparent: index != 2,
                    trailing: CupertinoButton(
                      padding: EdgeInsets.zero,
                      onPressed: widget.onBrightnessToggle,
                      child: Icon(
                        widget.brightness == Brightness.light
                            ? CupertinoIcons.moon_stars
                            : CupertinoIcons.sun_max,
                      ),
                    ),
                  ),
                  SliverSafeArea(
                    top: false,
                    sliver: SliverList.list(
                      children: [
                        CupertinoButton(
                          child: const Text('Next page'),
                          onPressed: () {
                            Navigator.of(context).push(
                              CupertinoPageRoute<void>(
                                builder: (BuildContext context) {
                                  return CupertinoPageScaffold(
                                    navigationBar: CupertinoNavigationBar(
                                      middle: Text('Inner page of tab $index'),
                                    ),
                                    child: ListView(
                                      children: [
                                        Center(
                                          child: CupertinoButton(
                                            child: const Text('Back'),
                                            onPressed: () {
                                              Navigator.of(context).pop();
                                            },
                                          ),
                                        ),
                                        if (index == 0) const _LongList(),
                                        const SizedBox(height: 20),
                                      ],
                                    ),
                                  );
                                },
                              ),
                            );
                          },
                        ),
                        if (index == 1) const _LongList(),
                        const SizedBox(height: 20),
                      ],
                    ),
                  ),
                ],
              ),
            );
          },
        );
      },
    );
  }
}

class _LongList extends StatelessWidget {
  const _LongList();

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        for (int i = 0; i < 50; i++) ...[
          CupertinoListTile(
            leading: const Icon(CupertinoIcons.book),
            title: Text('Bookstore item $i'),
          ),
        ],
      ],
    );
  }
}
  ```
</details>

This is the continuation of https://github.com/flutter/flutter/pull/142439.

I tried to keep the simplest API possible, so it's only introducing a new `automaticBackgroundVisibility` boolean parameter. 

In the implementation I had to use the `CupertinoPageScaffold` background color to make it look transparent instead of a 0 opacity, because of issues with route transitions.

I used an `InheritedWidget` so the nav bar is always getting the right background color from the parent scaffold, whether it is overridden or not. It would probably be better to make the inherited widget private but we'd need to move all the nav bar code to the same file as the scaffold, so for now I've just hidden it from the export. Let me know if it is okay to do that.

This PR is not dealing with the bottom tab bar, because the same [straightforward approach](https://github.com/veloce/flutter/commit/dde8ec6dc766036f73183ef8d6dedfe7ef49f082) doesn't work here. The problem is that the scroll notification is sent only when the scroll view is created or updated, so it doesn't work if one pushes a screen and navigates back.

Issues:
- #78607 
- #60411 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

veloce committed:
Add an initiallyTransparent param to cupertino nav bar widgets

veloce committed:
Fixes cupertino nav bar tests

veloce committed:
Update CupertinoNavigationBar API sample

veloce committed:
Make the border not transparent if parent scaffold color is null

veloce committed:
Add cupertino transparent nav bars tests

veloce committed:
Tweak initiallyTransparent doc comment

Piinks commented:
Hey @veloce thanks for contributing! It looks like there are some failing checks here, can you take a look?

veloce committed:
Don't use Color.lerp if not necessary

veloce committed:
Do not attempt to show widget name if not mounted

veloce commented on packages/flutter/lib/src/cupertino/colors.dart:
```diff
@@ -1102,7 +1102,7 @@ class CupertinoDynamicColor extends Color with Diagnosticable {
     ];
 
-    return '${_debugLabel ?? objectRuntimeType(this, 'CupertinoDynamicColor')}(${xs.join(', ')}, resolved by: ${_debugResolveContext?.widget ?? "UNRESOLVED"})';
+    return '${_debugLabel ?? objectRuntimeType(this, 'CupertinoDynamicColor')}(${xs.join(', ')}, resolved by: ${_debugResolveContext?.mounted ?? false ? _debugResolveContext?.widget : "UNRESOLVED"})';
```
I made this change because there were errors in the `flutter_gallery` smoke tests:

```
When the exception was thrown, this was the stack:
#0      Element.widget (package:flutter/src/widgets/framework.dart:3478:31)
#1      CupertinoDynamicColor.toString (package:flutter/src/cupertino/colors.dart:1105:135)
#2      DiagnosticsProperty.valueToString (package:flutter/src/foundation/diagnostics.dart:2616:60)
#3      DiagnosticsProperty.toDescription (package:flutter/src/foundation/diagnostics.dart:2633:21)
...
#111    RenderObject.toStringDeep (package:flutter/src/rendering/object.dart:3948:12)
#112    smokeDemo (dev/integration_tests/flutter_gallery/test/smoke_test.dart:82:102)
```

I could not yet track it down to the exact cause. This was due a bunch of some defunct `StatelessElement`, so I believe this is because of the background color transition.
If I remove the `Color.lerp` in the nav_bar code, or if I pass the same color to both `a` and `b` of `Color.lerp` there are no more errors.

MitchellGoodwin replied:
This seems like a problematic breaking change. It could make existing tests break.

veloce replied:
I can understand why it could be an unwanted change: if it would hide potential errors that are not supposed to happen.

But making existing tests break, I don't think it would do that. If a test call `CupertinoDynamicColor.toString` on a `CupertinoDynamicColor` that has defunct `Element` as `_debugResolveContext`, it would fail anyway. Like in the smoke tests here.

Now the question is: are we hiding problematic behaviour by making this change? 

Using only this PR as context, I'm not so sure it is a problem. I have isolated precisely the widget responsible for the error, as I explained in my previous comment. It is the persistent nav bar widget (which creates a color widget where the error happens). A simple
```
    print('${DateTime.now()} _persistentNavBar ${'${objectRuntimeType(context, '<optimized out>')}#${shortHash(context)}'}');
```
in its `build` method show that when we scroll, the framework will remove the existing `Element` associated to the widget from the tree and create another one, about at the moment the background color transition of the nav bar starts.

That the framework is doing such things in the background seems normal to me and should be transparent to the developer, so in theory I shouldn't try to change the code to make the framework behave differently.

So my conclusion is that with this patch, we've hit an edge case that has not occurred yet, involving the flutter_gallery smoke test `toStringDeep` and the `CupertinoDynamicColor.toString` which tries to get the context on a defunct element. So this change seems harmless after all.

What do you think @MitchellGoodwin ?

MitchellGoodwin replied:
I'm still nervous that this test breakage is showing some weird behaviors. Is the `setState` on the observer causing too much of a rebuild on every scroll? That would be a difference from the other PR.

I'd like to track down exactly why this is failing now and wasn't before. I'm worried about introducing a weird hiccup to app scrolling because that has a large affect in how the app feels. @Piinks are you able to take a look?

veloce replied:
What I found is that if you replace in my PR:

```dart
    final Color effectiveBackgroundColor = initiallyTransparent
        ? Color.lerp(parentPageScaffoldBackgroundColor ?? backgroundColor, backgroundColor, shrinkAnimationValue) ?? backgroundColor
        : backgroundColor;
``` 

by 
```dart
    final Color effectiveBackgroundColor = initiallyTransparent
        ? Color.lerp(backgroundColor.withOpacity(0), backgroundColor, shrinkAnimationValue) ?? backgroundColor
        : backgroundColor;
```

in `_LargeTitleNavigationBarSliverDelegate.build()`, then the framework is not replacing the element. I don't know why but it seems it can manage to keep the element when both colors are the same and only the opacity changes during a transition, but not when the colors change.

This is the reason why the tests were passing in the previous PR.

cc @MitchellGoodwin and @Piinks

MitchellGoodwin replied:
Interesting. I wonder if it's because parentPageScaffoldBackgroundColor is being set to a `final` instead of a `const`, and it's relying on a value from outside the widget's scope, so it's triggering replacing the widget? Similarly, `withOpacity` is not a static function, but `.lerp` is, and the `withOpacity` in that `.lerp` uses a constant value.

So is setting the `parentPageScaffoldBackgroundColor` to a const possible, and does that work? If not, would going the route with
```
Color.lerp(backgroundColor.withOpacity(0), backgroundColor, shrinkAnimationValue)
```
work? Does content show up in the navbar if partially scrolled? Would a safe area need to be added?

veloce replied:
I finally found the cause why the widget is replaced: the `BackdropFilter`. Since the `_wrapWithBackground` creates a backdrop filter only if the background is full opaque, when we transition from the scaffold color to the bar color, a widget is returned with and then without the `BackdropFilter` causing the element to be replaced.

If I change the code to:
```dart
  return ClipRect(
    child: BackdropFilter(
      filter: backgroundColor.alpha == 0xFF ? ImageFilter.blur() : ImageFilter.blur(sigmaX: 10.0, sigmaY: 10.0),
      child: childWithBackground,
    ),
  );
```

Now we don't have the widget replaced.

Note that going with the `withOpacity(0)` route, in order to avoid the above change, is not possible, because there is a blur artifact visible during the initial state, making it mandatory to update the check to `if (backgroundColor.alpha == 0xFF || backgroundColor.alpha == 0)`.

What should I do now? My guess is that it is preferable to always create the `BackdropFilter`, even though there is a 0 blur (is there a better way?) and remove the change in `CupertinoDynamicColor.toString`; but I'm waiting for your input.

MitchellGoodwin replied:
Great catch!

So that blur has been around for as long as the navbar has been in Flutter, but it's not gospel. Doing some digging, it's to allow for the navbar to be made transparent and have content scroll under it. It could theoretically be used to do this logic now, but that's a lot of work (as these PRs show) and ideally we would be able to do this by default. I think we should be looking at whether we want that `BackdropFilter` at all anymore, or at least not use it when the opacity can change dynamically.

Taking a step back, and looking at how SwiftUI handles native navbar, this logic is controlled by [.toolbarBackground(_:for:)](https://developer.apple.com/documentation/swiftui/view/toolbarbackground(_:for:)-7lv0f). It can be applied to either the NavBar or the TabBar, and controls whether the background is hidden or not with an enum with three values.

It works like this, as compared to Flutter:

| Background Visibility | SwiftUI                                                                              | Flutter                                                                                                                                                                                                                                                 |
|-----------------------|--------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Visible               | Always shows the background Color                                                    | Currently default in Flutter                                                                                                                                                                                                                            |
| Hidden                | Always has the background Color as transparent                                       | Flutter allows this, and applies a blur to underlying content. Currently if the opacity changes from 1.0 it replaces the widget. We want to continue to allow doing any opacity value to customize the transparency for statically transparent NavBars. |
| Automatic             | Default in SwiftUI. Makes the background transparent until content scrolls under it. | Flutter does not support this by default, and what this PR is trying to add.                                                                                                                                                                            |

I think we want to support all of this behavior. We could add an enum like Swift does, but my opinion is that it's better to add a bool property that turns off automatically changing the transparency of the navbar (it should change transparency by default). And if that is turned off, then we should do the blur like before. That would also make it easier to not swap in and out the BackdropFilter as opacity changes if the bool is true.

veloce replied:
Thanks for the detailed explanation!

If I understand correctly, flutter supports the "always transparent" behavior by allowing to explicitly pass a fully transparent background color. I think one of the known app that uses such an always hidden background is Signal.

I changed the boolean parameter name to reflect that we want to support all of this behavior. So it is now named `automaticBackgroundVisibility`.

It still feels like a workaround but I had to keep the parent scaffold background color to keep the route transitions looking as expected (don't have the time to figure out if it can be done differently).

I had to remove the tests related to the backdrop filter blur checks, because it always creates the `BackdropFilter` now. In theory we could avoid creating it if the `automaticBackgroundVisibility` is false AND the background fully opaque. I felt it was more confusing than useful to do that.

I will post new demos showcasing the 3 versions in this table.

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -432,17 +449,83 @@ class CupertinoNavigationBar extends StatefulWidget implements ObstructingPrefer
+    final Border? effectiveBorder = widget.border == null ? null : Border.lerp(initialBorder, widget.border, _scrollAnimationValue,);
+
+    final Color effectiveBackgroundColor = widget.initiallyTransparent
+      ? Color.lerp(parentPageScaffoldBackgroundColor ?? backgroundColor, backgroundColor, _scrollAnimationValue)!
```
This `!` seems suspicious and would be my guess on what's causing the debug error. Does it work if you set a `??` fallback?

veloce replied:
I think I tried with the fallback and it didn't work. But this is tricky I may have missed something. Will try again tomorrow.

veloce replied:
I tried to look at the differences between https://github.com/flutter/flutter/pull/142439 (because the smoke test passed there) and this PR. And I couldn't find any explanation for why it fails here.

MitchellGoodwin reviewed (commented):
Thank you for putting this PR together, this seems tricky. I'd like it if we could figure out that `toString` error. That may be a sign of something being pretty wrong.

Also, would it be possible to post a screen recording of what it looks like when you go backwards from a transparent background, back to an opaque one?

veloce commented on packages/flutter/lib/cupertino.dart:
```diff
@@ -48,7 +48,7 @@ export 'src/cupertino/list_tile.dart';
 export 'src/cupertino/magnifier.dart';
 export 'src/cupertino/nav_bar.dart';
-export 'src/cupertino/page_scaffold.dart';
+export 'src/cupertino/page_scaffold.dart' hide CupertinoPageScaffoldBackgroundColor;
```
No other reason that I wasn't sure about introducing this widget to the public API.

MitchellGoodwin replied:
We can let it be exposed then. It does no harm, and that way we avoid "why can't I use this Class" comments later when somebody comes up with an odd use case. 😅

Piinks replied:
Yeah, exporting everything is discussed a bit in our style guide. If we don't want to expose something, a private class would be more appropriate - which also helps to decide whether or not it should be exposed. If it cannot be private, then folks will likely need or want to use it.

veloce replied:
My bad then, I'll expose it :)

veloce committed:
Don't hide CupertinoPageScaffoldBackgroundColor

veloce committed:
Remove new line

veloce commented:
> Also, would it be possible to post a screen recording of what it looks like when you go backwards from a transparent background, back to an opaque one?


https://github.com/flutter/flutter/assets/423393/3847f2b5-0dac-4d5e-aa6f-03c1d2893e30

veloce committed:
Add a fallback to Color.lerp (even though it cannot be null in theory)

veloce commented:
I still cannot explain or fix the problem found in `flutter_gallery` smoke tests.

But using some debug `print` I think I could find which color is the culprit. This is a `CupertinoDynamicColor` defined in `CupertinoNavigationBarBackButton`:
https://github.com/veloce/flutter/blob/12c6446a9a0bfeee22f786cbc1c99c097dbbe25a/packages/flutter/lib/src/cupertino/nav_bar.dart#L1584

This is a color of a `TextStyle`, so it looks consistent with the stack trace of the test:

```
Null check operator used on a null value

When the exception was thrown, this was the stack:
#0      Element.widget (package:flutter/src/widgets/framework.dart:3478:31)
#1      CupertinoDynamicColor.toString (package:flutter/src/cupertino/colors.dart:1109:135)
#2      DiagnosticsProperty.valueToString (package:flutter/src/foundation/diagnostics.dart:2616:60)
#3      DiagnosticsProperty.toDescription (package:flutter/src/foundation/diagnostics.dart:2633:21)
#4      TextTreeRenderer._debugRender (package:flutter/src/foundation/diagnostics.dart:1195:31)
#5      TextTreeRenderer.render (package:flutter/src/foundation/diagnostics.dart:1125:14)
#6      TextTreeRenderer._debugRender (package:flutter/src/foundation/diagnostics.dart:1311:39)
#7      TextTreeRenderer.render (package:flutter/src/foundation/diagnostics.dart:1125:14)
#8      TextTreeRenderer._debugRender (package:flutter/src/foundation/diagnostics.dart:1369:33)
#9      TextTreeRenderer.render (package:flutter/src/foundation/diagnostics.dart:1125:14)
#10     TextTreeRenderer._debugRender (package:flutter/src/foundation/diagnostics.dart:1369:33)
#11     TextTreeRenderer.render (package:flutter/src/foundation/diagnostics.dart:1125:14)
#12     TextTreeRenderer._debugRender (package:flutter/src/foundation/diagnostics.dart:1369:33)
...
```


Now in my logs we can see the first `CupertinoDynamicColor.toString` which fails is due to the `_PersistentNavigationBar` widget:
```
00:22 +0: Flutter Gallery app smoke test (variant: TargetPlatform.android)
2024-05-31 16:40:50.917689 persistent nav bar element StatelessElement#0682a
back button element: StatelessElement#1ce2d
2024-05-31 16:40:51.167297 persistent nav bar element StatelessElement#14175
back button element: StatelessElement#2f4b2
2024-05-31 16:40:51.179605 persistent nav bar element StatelessElement#14175
back button element: StatelessElement#2f4b2
2024-05-31 16:40:51.397133 persistent nav bar element StatelessElement#033cf
back button element: StatelessElement#b9407
2024-05-31 16:40:51.631287 persistent nav bar element StatelessElement#300ad
2024-05-31 16:40:51.688865 persistent nav bar element StatelessElement#300ad
2024-05-31 16:40:51.715107 persistent nav bar element StatelessElement#300ad
2024-05-31 16:40:51.724946 persistent nav bar element StatelessElement#300ad
2024-05-31 16:40:51.734721 persistent nav bar element StatelessElement#300ad
2024-05-31 16:40:51.740004 persistent nav bar element StatelessElement#300ad
2024-05-31 16:40:51.745808 persistent nav bar element StatelessElement#300ad
2024-05-31 16:40:51.754271 persistent nav bar element StatelessElement#4553b
2024-05-31 16:40:51.763677 persistent nav bar element StatelessElement#4553b
2024-05-31 16:40:51.769524 persistent nav bar element StatelessElement#4553b
2024-05-31 16:40:51.774510 persistent nav bar element StatelessElement#4553b
2024-05-31 16:40:51.779014 persistent nav bar element StatelessElement#4553b
2024-05-31 16:40:51.985613: CupertinoDynamicColor.toString context StatelessElement#300ad(DEFUNCT)(no widget)
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following _TypeError was thrown running a test:
Null check operator used on a null value
```

Considering the timestamps, it looks like this happens during a scroll. Suddenly the framework creates another element for the same widget, and then it fails.

It don't have more time to debug this, will resume next week. But I'm really not sure how to fix this actually.

MitchellGoodwin commented:
Going to take the time to thank you @veloce again for your work on this so far. This has been much more of a rabbit hole than expected, and I appreciate your patience.

veloce committed:
Revert "Do not attempt to show widget name if not mounted"

This reverts commit 6056f5cb657987e300670d57dfff726d338f15a8.

veloce committed:
Implement cupertino nav bar automaticBackgroundVisibility

veloce committed:
Fix cupertino nav bar route transition when content is scrolled under

veloce commented:
> Going to take the time to thank you @veloce again for your work on this so far. This has been much more of a rabbit hole than expected, and I appreciate your patience.

No problem. I also haven't expected it to be that complicated, but I understand that things have to be done properly.

veloce commented:
Here are new demos, using a slightly modified version of flutter gallery:

## Automatic background visibility

https://github.com/flutter/flutter/assets/423393/d44ae4ac-a33f-42c9-a471-30f86c1aaf3f

## Alway hidden background

For this one I had to increase the `sigma` values, otherwise the buttons in the nav may not be visible and it is less pretty. The problem is that some flickering is noticeable around the edges, due to the blur.

https://github.com/flutter/flutter/assets/423393/9281a13e-741c-400e-83dd-d220283d54cd

## Always visible background

https://github.com/flutter/flutter/assets/423393/c8bd50ab-9842-4406-b239-5dd56060113e

veloce commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -169,13 +178,11 @@ Widget _wrapWithBackground({
     child: BackdropFilter(
-      filter: ImageFilter.blur(sigmaX: 10.0, sigmaY: 10.0),
+      filter: backgroundColor.alpha == 0xFF || !isContentScrolledUnder
+        ? ImageFilter.blur()
```
I wrote that to show it works but we'd need an `ImageFilter.noOp()` or similar instead.

MitchellGoodwin replied:
It looks like in the examples when it's supposed to be opaque there's still a blur. Is this why?

veloce replied:
By examples you mean the new screen recordings I posted? 

In that case it's normal to see blur everywhere since it did not make an example with a fully opaque background. 

I can make one if needed.

MitchellGoodwin replied:
If you wouldn't mind.

veloce replied:
I posted another one in the main comment section. About the `ImageFilter.blur()`, it works as expected as it has no visual effect. So having an `ImageFilter.noOp()` is to make it explicit we want a no op here.

Regarding the new demos that I pushed, is the visual result matching what you were expected as behavior?

MitchellGoodwin replied:
Yes it does, thank you. It looks great.

I think this is more than sufficient for this PR. I'd wonder if a lerp on the Blur would be the smoothest way to do it in the future. Just thinking out loud.

jonahwilliams replied:
you should add an `enabled` flag to BackdropFilter widget like we have for ImageFilter:

https://api.flutter.dev/flutter/widgets/ImageFiltered/ImageFiltered.html

There is no such thing as a no-op filter, this will still compute a blur which can be expensive.

veloce commented:
Here is another screen recording with an opaque background this time. The `CupertinoSliverNavigationBar` is configured with the default automatic visibility, and the inner page is configured with an always visible background.

https://github.com/flutter/flutter/assets/423393/7df9f925-6668-484f-a216-08ff4bd8b94e

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -339,6 +347,19 @@ class CupertinoNavigationBar extends StatefulWidget implements ObstructingPrefer
   /// {@endtemplate}
   final Color? backgroundColor;
```
Can you add something explaining how this will interact with `automaticBackGroundVisibility`? The sample code may need to be updated too.

MitchellGoodwin commented on packages/flutter/test/cupertino/nav_bar_test.dart:
```diff
@@ -58,48 +58,6 @@ void main() {
-    expect(find.byType(CupertinoNavigationBar), paints..rect(color: background.darkColor));
-  });
-
-  testWidgets('Non-opaque background adds blur effects', (WidgetTester tester) async {
```
I think we want to make sure that non-opaque backgrounds, with the automatic transparency turned off still apply a blur. It looks like your tests cover the other cases. Since we always have a BackdropFilter, can we check `BackdropFilter.filter`? And see if it's not a Blur with no sigma. At worst with `toString`?

MitchellGoodwin reviewed (commented):
Code LGTM. I think we want to make sure the original cases are checked in the tests, and that our documentation is up to date. The sample code for `CupertinoNavigationBar` also probably needs to be updated with the flag turned off, and maybe a new sample with the flag turned on should be added to show the automatic transparency? Or that original sample could be setup with a commented out line to turn it on and off and see the difference.

veloce committed:
Explain how nav bar background interacts with automaticBackgroundVisibility

veloce committed:
Add tests for cupertino nav bar blur effects

veloce commented:
I updated the doc comment and pushed new tests for the blur effects.

The sample was already updated (I changed the column for a `ListView` and increased the height of the coloured strips to be able to scroll up).

We can't keep the previous sample as is, (even with `automaticBackgroundVisibility` to false) because now the _wrapWithBackground function is changed to only add the blur if `isContentScrolledUnder` is true so it needs a `ScrollView`.

MitchellGoodwin commented:
> The sample was already updated (I changed the column for a ListView and increased the height of the coloured strips to be able to scroll up).

I saw that. I meant the sample should be able to show both the previous behavior and this new one. Maybe just adding a line setting the flag to true and a comment saying to change it to false to either show a static opaque or blurred navbar.

> We can't keep the previous sample as is, (even with automaticBackgroundVisibility to false) because now the _wrapWithBackground function is changed to only add the blur if isContentScrolledUnder is true so it needs a ScrollView.

So does that mean the original behavior wouldn't be possible unless content scrolls under? If somebody had a background image for their page and wanted the NavBar to always be blurred and partially show the image no matter if there's a scroll or not, would that be possible?

veloce commented:
> So does that mean the original behavior wouldn't be possible unless content scrolls under? 

In the current state, yes. 

> If somebody had a background image for their page and wanted the NavBar to always be blurred and partially show the image no matter if there's a scroll or not, would that be possible?

I think that's possible if I revert the blur logic to the original behaviour, which is to blur as long as the background is not fully opaque. The only drawback that I see is that it could lead to unwanted visual effects if the scaffold background color itself is semi-transparent (I suppose it is a very edge case).

veloce committed:
Add `enabled` flag to BackdropFilter, use it in cupertino nav bar

veloce committed:
Restore the original cupertino nav bar sample

veloce committed:
Fix cupertino example strips height

veloce committed:
Fix outdated comment

veloce committed:
Remove useless parameter

veloce committed:
Remove unused import

MitchellGoodwin commented:
> The only drawback that I see is that it could lead to unwanted visual effects if the scaffold background color itself is semi-transparent (I suppose it is a very edge case).

I think we will just have to leave that case up to developers to handle. Only so much we can cover.

MitchellGoodwin commented:
This breaks a fair amount of goldens, but that's to be expected. Cases where the navbar was sitting at the top of a scroll view are now transparent instead of having a background. Google testing will likely break for similar reasons, but that's on us.

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -432,17 +450,83 @@ class CupertinoNavigationBar extends StatefulWidget implements ObstructingPrefer
     final Color backgroundColor =
       CupertinoDynamicColor.maybeResolve(widget.backgroundColor, context) ?? CupertinoTheme.of(context).barBackgroundColor;
 
+    final Color? parentPageScaffoldBackgroundColor = CupertinoPageScaffoldBackgroundColor.maybeOf(context);
```
Thinking out loud, I think this is fine for this PR:

I wonder if there's a way to get a decent color in the case that there is no `CupertinoPageScaffoldBackgroundColor`. It's a little bit of an anti-pattern for us to rely on another widget in the tree for us, but navbars working with scaffolds is expected, and theoretically `CupertinoPageScaffoldBackgroundColor` could be used for other widgets.

MitchellGoodwin reviewed (approved):
LGTM! Let me get a second reviewer/ second opinion. Thank you for this contribustion. Excited to get this in.

MaherSafadii commented:
Finally, excited for this PR to get merged lol, I've been checking on it for the past month, thanks for your efforts @veloce

justinmc commented on packages/flutter/test/cupertino/nav_bar_test.dart:
```diff
@@ -1018,6 +1077,241 @@ void main() {
+      final BorderSide side = decoration.border!.bottom;
+      expect(side.color.opacity, 0.0);
+
+      // appears transparent since the background color is the same as the scaffold.
```
Capital A here.

justinmc commented on packages/flutter/test/cupertino/nav_bar_test.dart:
```diff
@@ -1018,6 +1077,241 @@ void main() {
+      final BorderSide side = decoration.border!.bottom;
+      expect(side.color.opacity, 0.0);
+
+      // appears transparent since the background color is the same as the scaffold.
```
Another captial A needed here.

justinmc reviewed (approved):
LGTM with nits 👍 . Thanks for taking the initiative to submit this PR! It looks very slick in the videos.

veloce commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -31,6 +31,10 @@ const double _kNavBarLargeTitleHeightExtension = 52.0;
 
+/// Number of logical pixels scrolled during which the navigation bar's background
+/// fades in or out.
+const double _kNavBarScrollUnderAnimationExtent = 10.0;
```
If eyeballing means trying to make it look as close as the native then yes; I empirically found this value worked well after trying several ones at random.

veloce committed:
Fixes after comments

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -432,17 +450,83 @@ class CupertinoNavigationBar extends StatefulWidget implements ObstructingPrefer
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _scrollNotificationObserver?.removeListener(_handleScrollNotification);
+    _scrollNotificationObserver = ScrollNotificationObserver.maybeOf(context);
+    _scrollNotificationObserver?.addListener(_handleScrollNotification);
```
I've been talking with some people about coordinating nav bar animations with scrolls. This is fine for now, but we need to find a better way to coordinate between the nav bar and the scrollable content. Animations are a frame behind, and there's issues with scrolling within nested navigation, multiple scrollable lists on a page, and ad to app scenarios. So good for now, ideally we will come up with a better communication channel and use that later.

MitchellGoodwin commented:
@veloce Could you please resolve the merge conflicts? Then we'll triage the golden tests and get this merged.

victorsanni commented:
Hi @veloce, issue #78607 is linked in this PR. Does this PR fix that issue?
